NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator.cpp

```diff
@@ -137,60 +178,111 @@ void LocalLinearLeastSquaresExtrapolator::extrapolateElement(
+        _nodal_values->add(indices, nodal_values.data());
+        counts.add(indices, std::vector<double>(indices.size(), 1.0));
+    }
 }
 
 void LocalLinearLeastSquaresExtrapolator::calculateResidualElement(
     std::size_t const element_index,
-    ExtrapolatableElementCollection const& extrapolatables)
+    const unsigned num_components,
+    ExtrapolatableElementCollection const& extrapolatables,
+    const double t,
+    GlobalVector const& current_solution,
+    LocalToGlobalIndexMap const& dof_table)
 {
     auto const& int_pt_vals = extrapolatables.getIntegrationPointValues(
-        element_index, _integration_point_values_cache);
+        element_index, t, current_solution, dof_table,
+        _integration_point_values_cache);
 
-    // TODO: for now always zeroth component is used
-    const auto& global_indices = _local_to_global(element_index, 0).rows;
+    auto const num_values = static_cast<unsigned>(int_pt_vals.size());
+    if (num_values % num_components != 0)
+        OGS_FATAL(
+            "The number of computed integration point values is not divisable "
+            "by the number of num_components. Maybe the computed property is "
+            "not a %d-component vector for each integration point.",
+            num_components);
 
-    const unsigned num_int_pts = int_pt_vals.size();
-    const unsigned num_nodes = global_indices.size();
+    // number of integration points in the element
+    const auto num_int_pts = num_values / num_components;
 
-    // filter nodal values of the current element
-    Eigen::VectorXd nodal_vals_element(num_nodes);
-    for (unsigned i = 0; i < num_nodes; ++i) {
-        // TODO PETSc negative indices?
-        nodal_vals_element[i] = _nodal_values[global_indices[i]];
-    }
+    const auto& global_indices =
+        _dof_table_single_component(element_index, 0).rows;
+    const auto num_nodes = static_cast<unsigned>(global_indices.size());
 
     auto const& interpolation_matrix =
         _qr_decomposition_cache.find({num_nodes, num_int_pts})->second.A;
 
-    double const residual = (interpolation_matrix * nodal_vals_element -
-                             MathLib::toVector(int_pt_vals))
-                                .squaredNorm();
+    Eigen::VectorXd nodal_vals_element(num_nodes);
+    auto const int_pt_vals_mat =
+        MathLib::toMatrix(int_pt_vals, num_components, num_int_pts);
+
+    for (unsigned comp = 0; comp < num_components; ++comp) {
+        // filter nodal values of the current element
+        for (unsigned i = 0; i < num_nodes; ++i) {
+            // TODO PETSc negative indices?
+            auto const idx = num_components * global_indices[i] + comp;
+            nodal_vals_element[i] = (*_nodal_values)[idx];
+        }
+
+        double const residual = (interpolation_matrix * nodal_vals_element -
+                                 int_pt_vals_mat.row(comp).transpose())
+                                    .squaredNorm();
```

Comment:
This residual computation and the interpolation procedure is documented somewhere, isn't it?